extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergCatalogAdapter.java

```diff
@@ -826,8 +798,500 @@ private Table readTableInternal(
+     * @param instructions The instructions for customizations while writing, or null to use default instructions
+     */
+    public void overwrite(
+            @NotNull final String tableIdentifier,
+            @NotNull final Table dhTable,
+            @Nullable final IcebergWriteInstructions instructions) {
+        overwrite(tableIdentifier, new Table[] {dhTable}, instructions);
+    }
+
+    /**
+     * Overwrite the existing iceberg table with the provided deephaven table in a single snapshot. This will change the
+     * schema of the existing table to match the provided deephaven table.
+     *
+     * @param tableIdentifier The identifier for the iceberg table to overwrite
+     * @param dhTable The deephaven table to overwrite with
+     * @param instructions The instructions for customizations while writing, or null to use default instructions
+     */
+    public void overwrite(
+            @NotNull final TableIdentifier tableIdentifier,
+            @NotNull final Table dhTable,
+            @Nullable final IcebergWriteInstructions instructions) {
+        overwrite(tableIdentifier, new Table[] {dhTable}, instructions);
+    }
+
+    /**
+     * Overwrite the existing iceberg table with the provided deephaven tables appended together in a single snapshot.
+     * All tables should have the same definition, else a table definition should be provided in the instructions. This
+     * will change the schema of the existing table to match the provided deephaven tables.
+     *
+     * @param tableIdentifier The identifier string for the iceberg table to overwrite
+     * @param dhTables The deephaven tables to overwrite with
+     * @param instructions The instructions for customizations while writing, or null to use default instructions
+     */
+    public void overwrite(
+            @NotNull final String tableIdentifier,
+            @NotNull final Table[] dhTables,
+            @Nullable final IcebergWriteInstructions instructions) {
+        overwrite(TableIdentifier.parse(tableIdentifier), dhTables, instructions);
+    }
+
+    /**
+     * Overwrite the existing iceberg table with the provided deephaven tables appended together in a single snapshot.
+     * All tables should have the same definition, else a table definition should be provided in the instructions. This
+     * will change the schema of the existing table to match the provided deephaven tables.
+     *
+     * @param tableIdentifier The identifier for the iceberg table to overwrite
+     * @param dhTables The deephaven tables to overwrite with
+     * @param instructions The instructions for customizations while writing, or null to use default instructions
+     */
+    public void overwrite(
+            @NotNull final TableIdentifier tableIdentifier,
+            @NotNull final Table[] dhTables,
+            @Nullable final IcebergWriteInstructions instructions) {
+        writeImpl(tableIdentifier, dhTables, instructions, true, true);
+    }
```

Comment:
I think we may want to consider ways to reduce the overload sprawl; ideally, there would be one entry point for append and one for overwrite - right now there are 4 append and 4 overwrite methods. We could consider a builder argument for the arguments.

I'm assuming after the ticking PR lands, we'll change this so that it's on the TableAdapter instead of CatalogAdapter.

Reply:
Yes, after the ticking PR lands, this will move into TableAdapter. 
After that, with the current layout, we will have two entry points for append, overwrite, etc.
```
public void append(Table dhTable, IcebergWriteInstructions instructions);
public void append(Table[] dhTables, IcebergWriteInstructions instructions);
```

This is very similar to our current parquet writing APIs. Also, I didn't want to move the table argument inside the instructions, since its the primary arguments to the API.

One alternate can be 
```
public void append(IcebergWriteInstructions instructions, Table... dhTables);
```
What do you think of that?

Reply:
If I do something like this, then I would have to replicate for all operations for the catalog adapter, including the read operations, which might be a bigger change that what we might want right now

Also, I can use `List` in existing APIs too:
```
public void append(List<Table> dhTable, IcebergWriteInstructions instructions);
```
The idea for having two different APIs is that if a user is writing one table, then they should not need to make a list.

Reply:
So, basically, you're suggesting that the API revolve around immutable parameter structs with builders? It's probably marginally more annoying to users in the console, but let's us reduce overload spam.

We could also just get out of the business of taking POJO table identifiers (or Strings; standardize on one or the other) cutting overloads by half. (Shivam points out that this will be automatically addressed by interposing the TableAdapter layer.)

`Table` args can just be a varargs list at the end of the method, cutting overloads again by half.